[GR-38416] Implement debug location info

adinn opened the pull request:
Implementation PR for  issue #4504

This includes location info for parameters and locals that ar ein registers or on the stack.

It does not yet include info for bindings to constants.

Updates to the reference manual will be provided as a separate PR.

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLocSectionImpl.java:
```diff
@@ -326,6 +334,56 @@ private int writeStackLocation(DebugContext context, int offset, byte[] buffer,
+                long l =  (kind == JavaKind.Long ? constant.asLong() : Double.doubleToRawLongBits(constant.asDouble()));
+                pos += putLong(l, scratch, 0);
+            }
+        } else {
```
If we would patch all the `put*` methods in `com.oracle.objectfile.elf.dwarf.DwarfSectionImpl` with something like
```patch
--- a/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfSectionImpl.java	(revision Staged)
+++ b/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfSectionImpl.java	(date 1651057782970)
@@ -147,10 +147,12 @@
      * Base level put methods that assume a non-null buffer.
      */
 
-    protected int putByte(byte b, byte[] buffer, int p) {
-        int pos = p;
+    protected int putByte(byte b, byte[] buf, int p) {
+        boolean useScratch = buf == null;
+        int pos = useScratch ? 0 : p;
+        byte[] buffer = useScratch ? scratch : buf;
         buffer[pos++] = b;
-        return pos;
+        return pos + (useScratch ? p : 0);
     }
 
     protected int putShort(short s, byte[] buffer, int p) {
```
we could get collapse these almost identical cases for `buffer == null` and `buffer != null`

olpaw replied:
> Yes, I agree that would be a very good idea, although I would prefer to do it in a follow-up PR.

Fine with me. I created https://github.com/oracle/graal/issues/4513 for that.

olpaw commented on substratevm/mx.substratevm/testhello.py:
```diff
@@ -438,40 +464,20 @@ def test():
-
-        checker.check(exec_string, skip_fails=True)
+    exec_string = execute("info args")
+    # we canto be sure whether "this" or the argument available
```
s/canto/want to ?

adinn replied:
Fixed

olpaw commented on substratevm/src/com.oracle.svm.test/src/hello/Hello.java:
```diff
@@ -178,4 +180,28 @@ private static void inlineTailRecursion(int n) {
+
+    @NeverInline("For testing purposes")
+    private static void noInlineManyArgs(int i0, int i1, int i2, int i3, int i4, int i5, int i6, int i7, int i8,
+                    float f0, float f1, float f2, float f3, float f4, float f5, float f6, float f7, float f8) {
```
What would currently happen with other primitive types like `boolean` or `double`?

adinn replied:
Nothing exciting I believe. All arguments occupy a register or an 8 byte stack slot.However, I could (or, more correctly, should) throw in some extra arguments of those types to check this works ok.

olpaw replied:
Yes please add other constants of primitive types here as well.

adinn replied:
Fixed

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfDebugInfo.java:
```diff
@@ -753,4 +777,77 @@ public int getAbstractInlineMethodIndex(ClassEntry classEntry, String methodName
+     * inline.
+     */
+
+    static final class DwarfLocalProperties {
```
Would it make sense to simply add an `int index` field in `NativeImageDebugLocalValueInfo` instead of managing those indexes externally?

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/Range.java:
```diff
@@ -245,110 +313,93 @@ public Range getSiblingCallee() {
+                        varRanges.add(subRange);
+                        break;
+                    case UNDEFINED:
+                        break;
```
```patch
--- a/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/Range.java	(revision Staged)
+++ b/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/Range.java	(date 1651055443390)
@@ -377,7 +377,7 @@
         for (int i = 0; i < localValueCount; i++) {
             DebugLocalValueInfo localValueInfo = subRange.getLocalValue(i);
             DebugLocalInfo local = subRange.getLocal(i);
-            if (local != null && localValueInfo.localKind() != DebugLocalValueInfo.LocalKind.UNDEFINED) {
+            if (local != null ) {
                 switch (localValueInfo.localKind()) {
                     case REGISTER:
                     case STACKSLOT:
```

adinn replied:
Yes, of course.

adinn replied:
Fixed

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLocSectionImpl.java:
```diff
@@ -0,0 +1,718 @@
+                return false;
+            }
+            // values need to be for the same line
+            if (value.line() != otherValue.line()) {
```
I'm curious, does it really happen that when we have `hi == otherLo` that `value.line() != otherValue.line()` is ever true?

adinn replied:
Yes, it can happen. The line number here is not for the declaration of the local var but for the infopoint where a location was recorded for the local. So, we can have two successive infopoints with different line numbers that both provide values for the local.

However, your question raises a good point all the same. It is important not to merge Range instances with different lines because the Range objects are used to drive generation of debug line info. However, I am now wondering whether we need to care about the line difference when merging location extents. I don't think the debugger lookup for locations requires that the location ranges conform to the line number ranges. So, I think we can remove this check and rely on the subsequent value comparison to decide whether or not to merge  the location extents.

adinn replied:
This got folded into the equals method where we do have to test for equality of line numbers.

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLocSectionImpl.java:
```diff
@@ -0,0 +1,718 @@
+                return false;
+            }
+            // locations must match
+            switch (value.localKind()) {
```
Shouldn't all those checks on DebugLocalValueInfos be implemented by providing an `public boolean equals(Object obj)` on `NativeImageDebugLocalValueInfo`. Also there seems to be some redundancy between here and `com.oracle.svm.hosted.image.NativeImageDebugInfoProvider.NativeImageDebugLocalValueInfo#sameAs`

adinn replied:
Yes, I think so. However, note that the above check is happening in the generator code. So, as well as implementing `boolean equals(Object obj)` in class `NativeImageDebugLocalValueInfo` we would also need to expose it as a method of interface `DebugLocalValueInfo`.

olpaw replied:
> we would also need to expose it as a method of interface DebugLocalValueInfo

Not sure why. Every Java object has `boolean equals(Object obj)` implicitly as part of its API. It's just that the default implementation is based of ref-equality instead of contents-equality. If you change this it an `equals` override in NativeImageDebugLocalValueInfo please also add `hashcode` for completeness (they should always come in pairs).

adinn replied:
Doh! ... :-)

adinn replied:
Fixed.

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLocSectionImpl.java:
```diff
@@ -0,0 +1,718 @@
+    private static final int DWARF_REG_X86_64_R15 = 15;
+    private static final int DWARF_REG_X86_64_XMM0 = 17;
+
+    private static final int[] GRAAL_X86_64_TO_DWARF_REG_MAP = {
```
Same as https://github.com/oracle/graal/pull/4505/files#r859968790

adinn commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLocSectionImpl.java:
```diff
@@ -0,0 +1,718 @@
+        JavaKind kind = constant.getJavaKind();
+        int byteCount = (kind == JavaKind.Object ? 8 : kind.getByteCount());
+        assert (kind != JavaKind.Object || constant.isNull()) : "only expecting null object constant!";
+        if (buffer == null) {
```
This needs fixing. The DWARF expression defining the constant needs to be preceded with a total byte count for the expression bytes. I accidentally deleted that before pushing the update for CONSTANT support.

adinn replied:
I corrected the error and also added support for writing object constants. So, you can now see constant strings in the initial heap etc.

olpaw commented:
Testing with `svmbuild/hello.hello`shows that setting a breakpoint in static method `java.lang.Integer#stringSize` does not seem to work.  If fact this method is not even found with `info functions stringSize`. If I set a line breakpoint in it I do see something but it looks buggy:
```
Thread 1 "hello.hello" hit Breakpoint 1, hello.Hello::noInlineManyArgs(int, int, int, int, int, int, int, int, int, float, float, float, float, float, float, float, float, float) (i0=0, i1=1, i2=2, i3=3, i4=4, i5=5, i6=6, i7=7, i8=8, f0=0, f1=1.125, f2=2.25, f3=3.375, f4=4.5, f5=5.625, f6=6.75, f7=7.875, f8=9) at hello/Hello.java:187
187	        System.out.println("i0 = " + i0);
(gdb) s
java.lang.StringConcatHelper::mix(long, int) (lengthCoder=5, value=0) at java/lang/StringConcatHelper.java:112
112	        return checkOverflow(lengthCoder + Integer.stringSize(value));
(gdb) 
java.lang.Integer::stringSize (x=0) at java/lang/Integer.java:551
551	        int d = 1;
(gdb) info locals
d = <optimized out>
p = <error reading variable p (Unhandled dwarf expression opcode 0xff)>
i = <optimized out>
```
Here are the remaining results of `info locals` within `stringSize`
```
552	        if (x >= 0) {
d = <error reading variable d (Unhandled dwarf expression opcode 0x1)>
p = <error reading variable p (Unhandled dwarf expression opcode 0xff)>
i = <optimized out>
(gdb) 
553	            d = 0;
d = <optimized out>
p = <error reading variable p (Unhandled dwarf expression opcode 0xff)>
i = <optimized out>
(gdb) 
554	            x = -x;
d = <optimized out>
p = <error reading variable p (Unhandled dwarf expression opcode 0xff)>
i = <optimized out>
(gdb) 
556	        int p = -10;
d = <optimized out>
p = <error reading variable p (Unhandled dwarf expression opcode 0xff)>
i = <optimized out>
(gdb) 
557	        for (int i = 1; i < 10; i++) {
d = <optimized out>
p = <error reading variable p (Asked for position 0 of stack, stack only has 0 elements on it.)>
i = <optimized out>
(gdb) 
558	            if (x > p)
d = <optimized out>
p = <error reading variable p (Asked for position 0 of stack, stack only has 0 elements on it.)>
i = <error reading variable i (Unhandled dwarf expression opcode 0x1)>
(gdb) 
559	                return i + d;
d = <optimized out>
p = <error reading variable p (Unhandled dwarf expression opcode 0xff)>
i = <optimized out>
```

olpaw commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java:
```diff
@@ -132,6 +153,7 @@ class NativeImageDebugInfoProvider implements DebugInfoProvider {
                         .flatMap(m -> Arrays.stream(m.getImplementations())
                                         .filter(Predicate.not(m::equals)))
                         .collect(Collectors.toSet());
+        wordBaseType = heap.getUniverse().getTypes().stream().filter(type -> type.getName().equals(GRAAL_WORDBASE_TYPE_NAME)).findFirst().get();
```
If you use
```patch
--- a/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImage.java	(revision Staged)
+++ b/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImage.java	(date 1651141290178)
@@ -471,7 +471,7 @@
                 Timer timer = TimerCollection.singleton().get(TimerCollection.Registry.DEBUG_INFO);
                 try (Timer.StopTimer t = timer.start()) {
                     ImageSingletons.add(SourceManager.class, new SourceManager());
-                    DebugInfoProvider provider = new NativeImageDebugInfoProvider(debug, codeCache, heap);
+                    DebugInfoProvider provider = new NativeImageDebugInfoProvider(debug, codeCache, heap, metaAccess);
                     objectFile.installDebugInfo(provider);
                 }
                 ProgressReporter.singleton().setDebugInfoTimer(timer);
```
then you have the HostedMetaAccess provider in the `NativeImageDebugInfoProvider` and you can get the `wordBaseType` via
```java
wordBaseType = metaAccess.lookupJavaType(WordBase.class);
```

olpaw replied:
It's a good idea in general to use metaAccess to get the HostedType equivalent of a given class.

adinn replied:
Fixed

olpaw reviewed (changes requested):
See review comments. Please also tidy up the commit history (remember commit message start with uppercase letter) and fold cleanup commits into their predecessor commits where it makes sense.

This was just the first round of reviewing. I will have another look mid next week.

adinn commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLocSectionImpl.java:
```diff
@@ -0,0 +1,718 @@
+    private static final int DWARF_REG_AARCH64_V31 = 95;
+
+    // map from compiler register indices to corresponding dwarf register
+    private static final int[] GRAAL_AARCH64_TO_DWARF_REG_MAP = {
```
Fixed

adinn commented:
@olpaw I addressed all the issues form your original review. While i was fixing the error I introduced into writing of primitive and null constants I also added the code for writing constant object values (i.e. reference to initial heap objects). So, you can now see things like String constants, albeit as references -- you have to execute `(gdb) print stringvar->value` to see the actual byte array contents.

I'll squash the commits down and check for proper commit messages next.

olpaw commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLocSectionImpl.java:
```diff
@@ -0,0 +1,756 @@
+    }
+
+    // map from compiler X86_64 register indices to corresponding dwarf AMD64 register index
+    private static final int[] GRAAL_X86_64_TO_DWARF_REG_MAP = {
```
Use the same pattern here as shown in https://github.com/oracle/graal/pull/4505/files#r865995904

adinn replied:
This doesn't actually work -- it does for AArch64 but only because that architecture has a more sensible naming scheme.

The AArch64 and X86_64 enum tags are sorted by the DWARF encoding. So, for x86_64 the register sequence for DWARF 0...7 is RAX(0),  RDX(1),  RCX(2), RBX(3), RSI(4), RDI(5), RBP(6), RSP(7).

The Graal register numbers do not follow the same order for their register  numbering. For numberings 0...7 the corresponding registers are RAX(0), RCX(1), RDX(2), RBX(3), RSP(4), RBP(5), RSI(6), RDI(7).

That different order is taken account of in the sequence in which the DwarfRegEncoding tags appear in the array initialization expression (notice how the first 8 entries in the following table are permuted). What it boils down to is that there are actually two translation steps at play in this iniitalization. The Graal register number n (range 0-31) indexes an array slot populated using a field expression that accesses the corresponding DwarfRegEncoding enum tag. The field retrieves the DWARF encoding for that DwarfRegEncoding.

This problem could be finessed by re-ordering the enum tags so that they are declared out of DWARF encoding order. If that were done then the iterative enumeration you are suggesting could be made to work. However, I think that would make this already fairly obscure translation process more difficult to follow.

I'd prefer to keep the current layout and add a comment to explain the rationale for this two specific step translation. As it stands, the code explicitly and independently displays each of the two steps in the translation via i) the argument passed to the enum tag and ii) the order in which entries appear in the initializer. Relying on the tag ordering to determine the permutation makes this step implicit and my view is that it would make the details of the mapping harder to understand and to document.

adinn replied:
I have pushed a patch that retains the current array initilaizer and documents the steps in the translation. I hope that is ok.

olpaw replied:
> Relying on the tag ordering to determine the permutation makes this step implicit and my view is that it would make the details of the mapping harder to understand and to document.

Alternatively you could just add another int field to the enum that holds the corresponding Graal register number explicitly for a given enum value (e.g. `regNumGraal`). Then again turning this into `GRAAL_X86_64_TO_DWARF_REG_MAP` is just a single line of code (similar to what I suggested before).

adinn replied:
Yes, that would be much better. I'll push another update.

adinn replied:
Ok, I did as you suggested.

n.b. To make this more bullet-proof I used the named Register instances {r0, ... v31} and {rax ... xmm15} from, respectively, AArch64 and AMD64 to specify the register numbers passed to the corresponding DwarfRegister enum tag. That required adding a dependency to the python suites exposing package jdk.interal.vm.ci/jdk.vm.ci.code to module com.oracle.objectfile . I hope that is ok.

adinn replied:
Repushed with format fixes.

olpaw commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java:
```diff
@@ -1211,79 +1729,497 @@ public int line() {
+        int stackParamCount;
+
+        ParamLocationProducer(ResolvedJavaMethod method) {
+            Architecture arch = ConfigurationValues.getTarget().arch;
```
Will this work across different OSes? If we have `OSX` or `Windows` instead of Linux?

adinn replied:
Doh! No, of course, this needs to employ an OS- and CPU-specific call protocol.

I actually wanted class ParamLocationProducer to employ the calling convention configuration data that drives the GraalVM compiler but I had no idea how to access it. So, instead I simply replicated the relevant info per CPU. Is there some simple way to obtain the compiler config as use it as an input to ParamLocationProducer?

For now I will fix this by multiplying cases in the implementation of ParamLocationProducer to respect both CPU and OS. However, it would be clearly be more 'future-proof' if the data that drives ParamLocationProducer was the same as that used by the compiler.

olpaw reviewed (changes requested):
PR is now very close to merging. There are only a few remaining issues.

adinn commented on substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLocSectionImpl.java:
```diff
@@ -0,0 +1,756 @@
+    }
+
+    // map from compiler AArch64 register indices to corresponding dwarf AArch64 register index
+    private static final int[] GRAAL_AARCH64_TO_DWARF_REG_MAP = {
```
I'm not sure this is the right thing to do. See below for details.

adinn replied:
Fixed by documenting the current code as per below.

olpaw commented:
@fniephaus this PR is now ready for internal CI.

fniephaus commented:
> @fniephaus this PR is now ready for internal CI.

 Great! I've opened an internal PR and am running the gates....

olpaw commented:
@adinn one of our internal CI gates found:
```
java.lang.AssertionError: Verification of jdk.vm.ci.meta.JavaMethod usage failed: Comparing 105|LoadField#method and 106|LoadField#method in HotSpotMethod<NativeImageDebugInfoProvider$NativeImageDebugLocationInfo.merge(NativeImageDebugInfoProvider$NativeImageDebugLocationInfo)> must use .equals() for object equality, not '==' or '!='
In com.oracle.svm.hosted.image.NativeImageDebugInfoProvider$NativeImageDebugLocationInfo.initLocalInfoList(NativeImageDebugInfoProvider.java:1646): parameter 4 of call to org.graalvm.compiler.debug.DebugContext.log(int, String, Object, Object) is a call to toString() which is redundant (the callee will do it) and forces unnecessary eager evaluation.
In com.oracle.svm.hosted.image.NativeImageDebugInfoProvider$NativeImageDebugLocationInfo.initSyntheticInfoList(NativeImageDebugInfoProvider.java:1679): parameter 4 of call to org.graalvm.compiler.debug.DebugContext.log(int, String, Object, Object) is a call to toString() which is redundant (the callee will do it) and forces unnecessary eager evaluation.
```

adinn commented:
> @adinn one of our internal CI gates found:
>
> > java.lang.AssertionError: Verification of jdk.vm.ci.meta.JavaMethod usage failed: Comparing 105|LoadField#method and 106|LoadField#method in HotSpotMethod<NativeImageDebugInfoProvider$NativeImageDebugLocationInfo.merge(NativeImageDebugInfoProvider$NativeImageDebugLocationInfo)> must use .equals() for object equality, not '==' or '!='
> > In com.oracle.svm.hosted.image.NativeImageDebugInfoProvider$NativeImageDebugLocationInfo.initLocalInfoList(NativeImageDebugInfoProvider.java:1646): parameter 4 of call to org.graalvm.compiler.debug.DebugContext.log(int, String, Object, Object) is a call to toString() which is redundant (the callee will do it) and forces unnecessary eager evaluation.
> > In com.oracle.svm.hosted.image.NativeImageDebugInfoProvider$NativeImageDebugLocationInfo.initSyntheticInfoList(NativeImageDebugInfoProvider.java:1679): parameter 4 of call to org.graalvm.compiler.debug.DebugContext.log(int, String, Object, Object) is a call to toString() which is redundant (the callee will do it) and forces unnecessary eager evaluation.

@olpaw I'll take a look at those errors and push a fix asap

teshull commented:
although it would probably be better to check `offsetsInSaveArea` instead of the list
